### ISSUE TITLE
hasManyJson saved as array of strings

### ISSUE THREAD

**Issue by carelloqwerty** [staudenmeir/eloquent-json-relations#24]
Project had been in use for a long time before I found the package, and all the many to many relations are stored as array of strings, something like {ids: ["2", "3", "4"]}. Relation "hasManyJson" in this case doesn't work and returns empty collection, it expects json field to be {ids: [2, 3, 4]}, but I must admit that "belongsToJson" method works perfectly. Also graphql schema requires ids to be a list of strings, so what solution can you advise to make 'hasManyJson' metod work regardless if he gets a list of strings or list of integers.
And also sync, attach and other methods  ruin everything, becouse they rewrite strings to integers in ids array and graphql fails.
To sum up: 
1) "belongsToJson" doesn't care if he gets list of integers or list of strings
2) "hasManyJson" requires only list of integers
it would be perfect if "hasManyJson" behaved same as "belongsToJson";

**Comment by staudenmeir** [comment]
The relations behave differently because they (have to) use different query features:
- `BelongsToJson` uses `WHERE IN` that can treat integer strings as integers.
- `HasManyJson` uses `WHERE JSON_CONTAINS()` that is strict about data types.

You can solve this issue by overriding and adjusting the `HasManyJson` relation in your application. I can help you with that, but I'm not going to include it into the package.

**Comment by carelloqwerty** [comment]
hm, seems it will be easier to write a migration and change all strings to int, becouse whereHas doesnt work either.
Inspecting queries I found that "whereHas" uses constraint jsonb_build_array to find all related entities that separate this array, while relation "HasManyJson" builds almost endless query "where ... or where ..." checking each id of collection. It might be good issue use same constraint with  jsonb_build_array for "HasManyJson"

**Comment by staudenmeir** [comment]
> Inspecting queries I found that "whereHas" uses constraint jsonb_build_array to find all related entities that separate this array, while relation "HasManyJson" builds almost endless query "where ... or where ..." checking each id of collection.

These are different kinds of queries and the `OR WHERE` clauses are all necessary.

You can't optimize eager loading queries with an array because you would only receive related records that contain *all* the parent keys. You need to receive records that contain *any* of the parent keys.

**Comment by carelloqwerty** [comment]
Oh, I've got it, in 'whereHas' we always use only id of parent entity, and after building array we get something like [ 1 ], and in this case contain _all_ and contain _any_ behaves the same.
All in all it this is amazing lib.
